packages/structures/src/channels/Channel.ts

```diff
@@ -0,0 +1,203 @@
+ * @typeParam Omitted - Specify the properties that will not be stored in the raw data field as a union, implement via `DataTemplate`
+ * @remarks Although this class _can_ be instantiated directly for any channel type,
+ * it's intended to be subclassed with the appropriate mixins for each channel type.
+ */
+export class Channel<
+	Type extends ChannelType | 'unknown' = ChannelType,
+	Omitted extends keyof ChannelDataType<Type> | '' = '',
+> extends Structure<ChannelDataType<Type>, Omitted> {
+	/**
+	 * The template used for removing data from the raw data stored for each Channel.
+	 *
+	 * @remarks This template is only guaranteed to apply to channels constructed directly via `new Channel()`.
+	 * Use the appropriate subclass template to remove data from that channel type.
+	 */
+	public static override DataTemplate: Partial<APIChannel> = {};
+
+	/**
+	 * @param data - The raw data received from the API for the channel
+	 */
+	public constructor(data: Partialize<ChannelDataType<Type>, Omitted>) {
+		super(data as ChannelDataType<Type>);
+	}
+
+	/**
+	 * {@inheritDoc Structure._patch}
+	 *
+	 * @internal
+	 */
+	public override _patch(data: Partial<ChannelDataType<Type>>) {
+		return super._patch(data);
+	}
+
+	/**
+	 * The id of the channel
+	 */
+	public get id() {
+		return this[kData].id;
+	}
+
+	/**
+	 * The type of the channel
+	 */
+	public get type() {
+		// This cast can be incorrect when type is omitted and if the wrong type of channel was constructed
+		return this[kData].type as Type extends 'unknown' ? number : Type;
+	}
+
+	/**
+	 * The name of the channel, null for DMs
+	 *
+	 * @privateRemarks The type of `name` can be narrowed in Guild Channels and DM channels to string and null respectively,
+	 * respecting Omit behaviors
+	 */
+	public get name() {
+		return this[kData].name;
+	}
+
+	/**
+	 * The flags that are applied to the channel.
+	 *
+	 * @privateRemarks The type of `flags` can be narrowed in Guild Channels and DMChannel to ChannelFlags, and in GroupDM channel
+	 * to null, respecting Omit behaviors
+	 */
+	public get flags() {
+		const flags =
+			'flags' in this[kData] && typeof this[kData].flags === 'number' ? (this[kData].flags as ChannelFlags) : null;
+		return flags ? new ChannelFlagsBitField(flags) : null;
+	}
+
+	/**
+	 * The timestamp the channel was created at
+	 */
+	public get createdTimestamp() {
+		return ['string', 'bigint'].includes(typeof this.id)
+			? DiscordSnowflake.timestampFrom(this.id as bigint | string)
+			: null;
+	}
+
+	/**
+	 * The time the channel was created at
+	 */
+	public get createdAt() {
+		const createdTimestamp = this.createdTimestamp;
+		return createdTimestamp ? new Date(createdTimestamp) : null;
+	}
+
+	/**
+	 * Indicates whether this channel is a thread channel
+	 *
+	 * @privateRemarks Overriden to `true` on `ThreadChannelMixin`
+	 */
+	public isThread(): this is ThreadChannelMixin<Extract<Type, ThreadChannelType>> {
```

Comment:
Shouldn't these assertions be `this is this & MixInClass`? Cause otherwise won't TS narrow the type to ONLY those properties?

Reply:
Good news: it's not an infinite recursion
bad news: with more than 5 of these typeguards v8 reaches its limits on my M2 with 8GB RAM

Reply:
mmm, not good
the typeguard in its current form is also not really acceptable, since I'd imagine it leads into weird cases like
```ts
if (!channel.isThread()) return;

// can't use the normal `Channel` props now?
```